File: frameworks/Java/kora/kora-loom-undertow/src/main/java/ru/tinkoff/kora/techempower/loom/undertow/TechempowerController.java
package ru.tinkoff.kora.techempower.loom.undertow;

import jakarta.annotation.Nullable;
import ru.tinkoff.kora.http.common.annotation.HttpRoute;
import ru.tinkoff.kora.http.common.annotation.Query;
import ru.tinkoff.kora.http.common.body.HttpBody;
import ru.tinkoff.kora.http.server.common.HttpServerResponse;
import ru.tinkoff.kora.http.server.common.annotation.HttpController;
import ru.tinkoff.kora.json.common.annotation.Json;
import ru.tinkoff.kora.jte.common.Jte;
import ru.tinkoff.kora.techempower.common.Fortune;
import ru.tinkoff.kora.techempower.common.JsonResponse;
import ru.tinkoff.kora.techempower.common.World;

import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

import static java.util.Comparator.comparing;

@HttpController
public final class TechempowerController {
    private final TechempowerRepository repository;

    public TechempowerController(TechempowerRepository repository) {
        this.repository = repository;
    }


    @HttpRoute(method = "GET", path = "/json")
    @Json
    public JsonResponse json() {
        return new JsonResponse("Hello, World!");
    }

    @HttpRoute(method = "GET", path = "/db")
    @Json
    public World db() {
        var id = World.randomWorldNumber();
        return this.repository.findWorldById(id);
    }

    @HttpRoute(method = "GET", path = "/queries")
    @Json
    public List<World> queries(@Query("queries") @Nullable String queriesStr) {
        int queries = World.parseQueryCount(queriesStr);
        return this.repository.getJdbcConnectionFactory().withConnection(() -> {
            var result = new ArrayList<World>(queries);
            for (int i = 0; i < queries; i++) {
                var id = World.randomWorldNumber();
                result.add(this.repository.findWorldById(id));
            }
            return result;
        });
    }

    @HttpRoute(method = "GET", path = "/fortunes")
    @Jte
    public FortunesTemplate fortunes() {
        var fortunes = this.repository.fortunes();
        fortunes.add(new Fortune(0, "Additional fortune added at request time."));
        fortunes.sort(comparing(Fortune::message));
        return new FortunesTemplate(fortunes);
    }

    @HttpRoute(method = "GET", path = "/updates")
    @Json
    public List<World> updates(@Query("queries") @Nullable String queriesStr) {
        int queries = World.parseQueryCount(queriesStr);
        var result = new ArrayList<World>(queries);
        this.repository.getJdbcConnectionFactory().withConnection(() -> {
            for (var i = 0; i < queries; i++) {
                var world = this.repository.findWorldById(World.randomWorldNumber());
                var newNumber = World.randomWorldNumber(world.randomNumber());
                result.add(new World(world.id(), newNumber));
            }
            result.sort(comparing(World::id));
            this.repository.update(result);
        });
        return result;
    }

    private static final ByteBuffer PLAINTEXT_RESPONSE = ByteBuffer.wrap("Hello, World!".getBytes(StandardCharsets.UTF_8));

    @HttpRoute(method = "GET", path = "/plaintext")
    public HttpServerResponse plaintext() {
        return HttpServerResponse.of(200, HttpBody.plaintext(PLAINTEXT_RESPONSE));
    }
}

File: frameworks/Java/kora/kora-kotlin/src/main/kotlin/ru/tinkoff/kora/techempower/kotlin/TechempowerController.kt
package ru.tinkoff.kora.techempower.kotlin

import kotlinx.coroutines.*
import kotlinx.coroutines.future.await
import kotlinx.coroutines.future.future
import ru.tinkoff.kora.common.Context
import ru.tinkoff.kora.database.vertx.VertxRepository
import ru.tinkoff.kora.http.common.HttpBody
import ru.tinkoff.kora.http.common.annotation.HttpRoute
import ru.tinkoff.kora.http.common.annotation.Query
import ru.tinkoff.kora.http.server.common.HttpServerResponse
import ru.tinkoff.kora.http.server.common.annotation.HttpController
import ru.tinkoff.kora.json.common.annotation.Json
import ru.tinkoff.kora.jte.common.Jte
import ru.tinkoff.kora.techempower.common.World.*
import java.nio.ByteBuffer
import java.nio.charset.StandardCharsets
import kotlin.coroutines.coroutineContext

@HttpController
class TechempowerController(private val repository: TechempowerRepository) {

    private val plaintextResponse = ByteBuffer.wrap("Hello, World!".toByteArray(StandardCharsets.UTF_8))

    @HttpRoute(method = "GET", path = "/plaintext")
    suspend fun plaintext() = HttpServerResponse.of(200, HttpBody.plaintext(plaintextResponse))!!

    @HttpRoute(method = "GET", path = "/json")
    @Json
    suspend fun json() = JsonResponse("Hello, World!")

    @HttpRoute(method = "GET", path = "/db")
    @Json
    suspend fun db(): World = repository.findWorldById(randomWorldNumber())

    @HttpRoute(method = "GET", path = "/queries")
    @Json
    suspend fun queries(@Query("queries") queriesStr: String?): List<World> {
        val queries = parseQueryCount(queriesStr)
        return repository.withConnection {
            val result = ArrayList<Deferred<World>>(queries)
            for (i in 0 until queries) {
                val id = randomWorldNumber()
                result.add(async { repository.findWorldById(id) })
            }
            result.map { it.await() }
        }
    }

    @HttpRoute(method = "GET", path = "/updates")
    @Json
    suspend fun update(@Query("queries") queriesStr: String?): List<World> {
        val queries = parseQueryCount(queriesStr)
        return repository.withConnection {
            val result = ArrayList<Deferred<World>>(queries)
            for (i in 0 until queries) {
                val id = randomWorldNumber()
                result.add(async { repository.findWorldById(id) })
            }
            val newWorlds = result.map {
                val world = it.await()
                World(world.id, randomWorldNumber(world.randomNumber))
            }
            repository.updateList(newWorlds)
            newWorlds
        }
    }

    @HttpRoute(method = "GET", path = "/fortunes")
    @Jte
    suspend fun fortunes(): FortunesTemplate {
        val fortunes = repository.fortunes() as MutableList<Fortune>
        fortunes.add(Fortune(0, "Additional fortune added at request time."))
        fortunes.sortBy { it.message }
        return FortunesTemplate(fortunes)
    }

    private suspend fun <T> VertxRepository.withConnection(callback: suspend CoroutineScope.() -> T): T {
        val repository = this
        val scope = CoroutineScope(coroutineContext + Dispatchers.Unconfined)
        val future = repository.vertxConnectionFactory.withConnectionCompletionStage {
            scope.future { callback(scope) }
        }
        return future.await()
    }

}
